flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1046,6 +1063,63 @@ class ScalarExprLowering {
       llvm_unreachable("unhandled constant");
     }
   }
+
+  /// Generate a raw literal value and store it in the rawVals vector.
+  template <Fortran::common::TypeCategory TC, int KIND>
+  void
+  genRawLit(const Fortran::evaluate::Scalar<Fortran::evaluate::Type<TC, KIND>>
+                &value) {
+    mlir::Attribute val;
+    if constexpr (TC == Fortran::common::TypeCategory::Integer) {
+      auto rTy = builder.getIntegerType(rawValWidth);
+      val = builder.getIntegerAttr(rTy, value.ToInt64());
+    } else if constexpr (TC == Fortran::common::TypeCategory::Logical) {
+      auto rTy = builder.getIntegerType(rawValWidth);
+      val = builder.getIntegerAttr(rTy, value.IsTrue());
+    } else if constexpr (TC == Fortran::common::TypeCategory::Real) {
+      std::string str = value.DumpHexadecimal();
+      auto rTy = genRealType(builder.getContext(), rawValWidth);
+      if constexpr (KIND == 2) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEhalf(), str};
```

Comment:
This all bypasses the KindMap. We don't hard-code kind values from the front end like this. See FIRBuilder.cpp and createRealZeroConstant, for example.

Reply:
I haven't had any luck incorporating your suggestion. As an FYI, I based the code I'm using on the genScalarLit() function in the same file.

Reply:
If you call `kindMap.getFloatSemantics(KIND)`, the first argument to `llvm::APFloat` constructor is returned. The APFloat can use a string representation. See FIRAttr.cpp:158.